Feature/objects 712 v2 column definitions

andreassiegel opened the pull request:
Updated the `ObjectEntity` to match the v2 entity as well as the column definitions.

Also replaced the `columnDefinition` to be consistent, and to let Hibernate do the column definition work.

Commit:
Fix column name case to match v2 names

Commit:
LastModified as long in converter

Commit:
Change dates to long in persistence service

Commit:
Update ObjectEntity to match v2 definitions

andreassiegel commented:
Ah, wait... There is an issue with creating objects due to missing default values.

Commit:
Remove unnecessary annotations or column annotation attributes: lastModifiedTimestamp may be null now

wfschneider-rfid commented on src/main/java/net/smartcosmos/dao/objects/domain/ObjectEntity.java:
```diff
@@ -24,43 +34,69 @@
     private UUID accountUrn;
 
     @CreatedDate
     @Column(name = "created", insertable = true, updatable = false)
```
name should be `createdTimestamp`

wfschneider-rfid commented on src/main/java/net/smartcosmos/dao/objects/domain/ObjectEntity.java:
```diff
@@ -24,43 +34,69 @@
     @Column(name = "created", insertable = true, updatable = false)
-    @Temporal(TemporalType.TIMESTAMP)
-    private Date created;
+    private Long created;
```
What is the benefit of a `Long` over a `DateTime` other than consistency with the past?  The converters can convert this to/from anything we want.  I think a `DateTime` makes more sense here.

andreassiegel replied:
For the future it makes more sense, right. However, staying with the `Long` allows for reuse of old databases.
If we go for a v2-like default implementation, it should IMO be as compatible as possible.

wfschneider-rfid commented on src/main/java/net/smartcosmos/dao/objects/domain/ObjectEntity.java:
```diff
@@ -24,43 +34,69 @@
+    @Basic
+    @NotNull
+    @Column(name="activeFlag", nullable = false)
     private Boolean activeFlag;
```
good Java naming calls this `isActive` in the class.  I would argue the boolean type negates the usefulness of the `flag` in the database as well.
```
@Column(name="active", nullable = false)
private Boolean isActive;
```

andreassiegel replied:
Lombok would call the getter `getIsActive()` then -- that's even more ugly. :/

wfschneider-rfid replied:
bad call on Lombok then

andreassiegel replied:
Lombok just creates getters like `isActive()` for `boolean` fields, but not for `Boolean`. Meh.
https://projectlombok.org/features/GetterSetter.html

Alternative: Override the Lombok getter: http://stackoverflow.com/questions/18139678/lombok-how-to-customise-getter-for-boolean-object-field

wfschneider-rfid commented on src/main/java/net/smartcosmos/dao/objects/domain/ObjectEntity.java:
```diff
@@ -24,43 +34,69 @@
-    @Column(name="activeflag")
+    @Basic
+    @NotNull
+    @Column(name="activeFlag", nullable = false)
```
What type column does this become?  I think we should we make it `bit(1)` and default to `true`.

andreassiegel replied:
It's a `bit(1)` column and defaults to `true`, but in v2 the default value was `false`.

Commit:
Return creation timestamp in ObjectResponse.lastModifiedTimestamp if ObjectEntity.lastModified is null
The converter solution keeps consistency with Objects v2 from the API perspective while also allowing for appropriate use of lastModified and created for future use cases. As Tim said, he already did something similar with the timestamps somewhere else.

wfschneider-rfid commented on src/main/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceService.java:
```diff
@@ -117,11 +113,11 @@ public ObjectResponse create(String accountUrn, ObjectCreate createObject) {
         // findByExample doesn't deal with dates, so we have to do it ourselves
-        Date modifiedAfterDate = new Date(0);
+        Long modifiedAfterDate = null;
         boolean modifiedAfterDateSpecified = false;
```
We can't be doing things like this.  There is no need to create and use a separate variable to represent the presence of something.  Initialize the `modifiedAfterDate` field to `null` as it is here and later check it for `!= null`.  One less variable.

wfschneider-rfid commented on src/main/java/net/smartcosmos/dao/objects/impl/ObjectPersistenceService.java:
```diff
@@ -133,11 +129,11 @@ public ObjectResponse create(String accountUrn, ObjectCreate createObject) {
             if (singleResultLastModified == null){
                 singleResultLastModified = singleResult.getCreated();
             }
-            if(!modifiedAfterDateSpecified || singleResultLastModified.after(modifiedAfterDate))
```
We cannot be doing this.  We need to make the database do the work by invoking the query with a comparison for the `lastUpdatedTimestamp`.  We should never retrieve results and iterate or inspect the return wen that can be most effectively done by the database.

Commit:
Rename created to createdTimestamp

voor committed:
Merge branch 'master' into feature/OBJECTS-712_v2_column_definitions

Commit:
Set lastModified already on create to be consistent with v2

As Robert mentioned yesterday, v2 and v3 might use the same database for the testing in the beginning, therefore we shouldn't want to have NULL values in this database, as long as we don't tell v2 how to handle those. That might not be perfect, but as Will said this JPA implementation will be the v2-compatibility version - not the actual v3 example implementation. The real v3 one could then use real timestamps instead of long, populate lastModified only on modifications etc.

Commit:
Merge branch 'feature/OBJECTS-712_v2_column_definitions' of github.com:SMARTRACTECHNOLOGY/smartcosmos-dao-objects-jpa into feature/OBJECTS-712_v2_column_definitions

Commit:
Remove modifiedAfterDateSpecified variable, use != null check instead

Commit:
Merge branch 'master' into feature/OBJECTS-712_v2_column_definitions

Commit:
lastModified not nullable

Commit:
Rename accountUuid to accountId

Commit:
Replace @AllArgsConstructor annotation for @Builder by explicit custom AllArgsConstructor taking care of initialization
